src/lib/tls/msg_client_hello.cpp

```diff
@@ -431,59 +505,98 @@ Client_Hello_12::Client_Hello_12(Handshake_IO& io,
 
    if(policy.support_cert_status_message())
-      m_extensions.add(new Certificate_Status_Request({}, {}));
+      m_data->extensions.add(new Certificate_Status_Request({}, {}));
 
    auto supported_groups = std::make_unique<Supported_Groups>(policy.key_exchange_groups());
 
    if(!supported_groups->ec_groups().empty())
       {
-      m_extensions.add(new Supported_Point_Formats(policy.use_ecc_point_compression()));
+      m_data->extensions.add(new Supported_Point_Formats(policy.use_ecc_point_compression()));
       }
 
-   m_extensions.add(supported_groups.release());
+   m_data->extensions.add(supported_groups.release());
 
-   m_extensions.add(new Signature_Algorithms(policy.acceptable_signature_schemes()));
+   m_data->extensions.add(new Signature_Algorithms(policy.acceptable_signature_schemes()));
 
    if(reneg_info.empty() && !next_protocols.empty())
-      { m_extensions.add(new Application_Layer_Protocol_Notification(next_protocols)); }
+      { m_data->extensions.add(new Application_Layer_Protocol_Notification(next_protocols)); }
 
-   cb.tls_modify_extensions(m_extensions, CLIENT, type());
+   cb.tls_modify_extensions(m_data->extensions, CLIENT, type());
 
    hash.update(io.send(*this));
    }
 
 #if defined(BOTAN_HAS_TLS_13)
 
+Client_Hello_13::Client_Hello_13(std::unique_ptr<Client_Hello_Internal> data)
+   : Client_Hello(std::move(data))
+   {
+   // RFC 8446 4.1.2
+   //    TLS 1.3 ClientHellos are identified as having a legacy_version of
+   //    0x0303 and a "supported_versions" extension present with 0x0304 as the
+   //    highest version indicated therein.
+   //
+   // Note that we already checked for "supported_versions" before entering this
+   // c'tor in `Client_Hello_13::parse()`. This is just to be doubly sure.
+   BOTAN_ASSERT_NOMSG(m_data->extensions.has<Supported_Versions>());
+
+   // RFC 8446 4.2.1
+   //    Servers MAY abort the handshake upon receiving a ClientHello with
+   //    legacy_version 0x0304 or later.
+   if(m_data->legacy_version != Protocol_Version::TLS_V12)
+      {
+      throw TLS_Exception(Alert::DECODE_ERROR,
+                          "TLS 1.3 Client Hello has invalid legacy_version");
+      }
+
+   // RFC 8446 4.1.2
+   //    For every TLS 1.3 ClientHello, [the compression method] MUST contain
+   //    exactly one byte, set to zero, [...].  If a TLS 1.3 ClientHello is
+   //    received with any other value in this field, the server MUST abort the
+   //    handshake with an "illegal_parameter" alert.
+   if(m_data->comp_methods.size() != 1 || m_data->comp_methods.front() != 0)
+      {
+      throw TLS_Exception(Alert::ILLEGAL_PARAMETER,
+                          "Client sent TLS 1.3 Client Hello with compression "
+                          "method != null");
+      }
+
+   // TODO: Reject oid_filters extension if found (which is the only known extension that
+   //       must not occur in the TLS 1.3 client hello.
```

Comment:
Should this be done in this PR or is it done in a followup in the stack? Either is fine, just checking.

Reply:
It'll be a follow-up.